wallet-new/src/Cardano/Wallet/Kernel/Actions.hs

```diff
@@ -0,0 +1,186 @@
+
+module Cardano.Wallet.Kernel.Actions
+  ( WalletAction(..)
+  , WalletActionInterp(..)
+  , forkWalletWorker
+  , walletWorker
+  , interp
+  , interpList
+  , WalletWorkerState
+  , isInitialState
+  , hasPendingFork
+  , isValidState
+  ) where
+
+import           Universum
+import           Control.Concurrent.Async (async, link)
+import           Control.Concurrent.Chan
+import           Control.Lens (makeLenses, (%=), (.=), (+=), (-=))
+import qualified Data.Text.Buildable
+import           Formatting (bprint, shown, build, (%))
+
+import           Pos.Util.Chrono
+
+{-------------------------------------------------------------------------------
+  Workers and helpers for performing wallet state updates
+-------------------------------------------------------------------------------}
+
+-- | Actions that can be invoked on a wallet, via a worker.
+--   Workers may not respond directly to each action; for example,
+--   a `RollbackBlocks` followed by several `ApplyBlocks` may be
+--   batched into a single operation on the actual wallet.
+data WalletAction b
+    = ApplyBlocks    (OldestFirst NE b)
+    | RollbackBlocks (NewestFirst NE b)
+    | LogMessage Text
+
+-- | Interface abstraction for the wallet worker.
+--   The caller provides these primitive wallet operations;
+--   the worker uses these to invoke changes to the
+--   underlying wallet.
+data WalletActionInterp m b = WalletActionInterp
+    { applyBlocks :: OldestFirst NE b -> m ()
+    , switchToFork :: Int -> OldestFirst [] b -> m ()
+    , emit :: Text -> m ()
+    }
+
+-- | Internal state of the wallet worker.
+data WalletWorkerState b = WalletWorkerState
+    { _pendingRollbacks    :: !Int
+    , _pendingBlocks       :: !(NewestFirst [] b)
+    , _lengthPendingBlocks :: !Int
+    }
+  deriving Eq
+
+makeLenses ''WalletWorkerState
+
+-- A helper function for lifting a `WalletActionInterp` through a monad transformer.
+lifted :: (Monad m, MonadTrans t) => WalletActionInterp m b -> WalletActionInterp (t m) b
+lifted i = WalletActionInterp
+    { applyBlocks  = lift . applyBlocks i
+    , switchToFork = \n bs -> lift (switchToFork i n bs)
+    , emit         = lift . emit i
+    }
```

Comment:
Given a suitable `hoist` function, you could write this as `hoist lift`

Reply:
Right. I mean something [like this](https://github.com/input-output-hk/cardano-sl/blob/develop/wallet-new/src/Cardano/Wallet/Client.hs#L181-L223). `hoist` is probably not the right word for it